### PR TITLE
Switch to pytest-mpl for image testing

### PR DESCRIPTION
**Pull request by QuLogic**
With `pytest-mpl`, we can throw away a bunch of code for image comparison and testing. I don't consider that public API, so didn't bother writing any deprecation.

To minimize changes, I returned `ax.figure` when we have an Axes, but no explicit Figure. Also, I added in a pytest hook to generate the baseline image directory in some manner that matches the old one.

### CODE REVIEW

**Comment by QuLogic** [comment]
You can see that failed uploads work here where I removed all non-default tolerances: https://github.com/QuLogic/cartopy/actions/runs/1268110201

File: lib/cartopy/tests/mpl/conftest.py [hunk 1]
```diff
@@ -19,3 +19,20 @@ def mpl_test_cleanup(request):
     finally:
         # Closes all open figures and switches backend back to original
         plt.switch_backend(orig_backend)
+
+
+def pytest_itemcollected(item):
+    mpl = item.get_closest_marker('mpl_image_compare')
```
Comment (greglucas) [inline lib/cartopy/tests/mpl/conftest.py:25]: Perhaps a different name than `mpl`? This confused me thinking it was `import matplotlib as mpl` down below.

**Review by greglucas — COMMENTED** [review]
Looks like we may need to think about the `PYPROJ_GLOBAL_CONTEXT` as there are locks on the PROJ database that occasionally cause the tests to fail.

**Review by greglucas — APPROVED** [review]

**Comment by QuLogic** [comment]
There used to be a lock on all image tests, but they were only supposed to prevent the 1 or 2 tests that used the same test image from clashing. Since `pytest-mpl` saves the test images in different directories, this shouldn't be needed. But it seems that the slowdown of the locking was enough to prevent Proj contexts from stomping on each other.

I don't know if pre-downloading the required grids will help (I already know which ones are needed from #1880).

There have been several threading/multiprocess bugs in both Proj and pyproj, and I'm not sure if conda-forge has all the fixes, or if they are all released. Maybe @snowman2 knows?

**Comment by snowman2** [comment]
> I'm not sure if conda-forge has all the fixes, or if they are all released

They are all released. You just can't use the global context if you need thread safety.

**Comment by QuLogic** [comment]
> > I'm not sure if conda-forge has all the fixes, or if they are all released
> 
> They are all released.

But on conda-forge too? We only get proj 8.0.1, looks like.

> You just can't use the global context if you need thread safety.

pytest-xdist is process based, I believe, not threaded.

**Comment by snowman2** [comment]
> pytest-xdist is process based, I believe, not threaded

I would be surprised if the global context was safe for processes.

**Comment by snowman2** [comment]
> But on conda-forge too? We only get proj 8.0.1, looks like.

It depends on the dependencies you have. If you only install phone pyproj, you will get the latest. Other dependencies might have a migration PR pending.

**Comment by snowman2** [comment]
I looked at the failure message and I think it is due to the fact that PROJ NETWORK is attempting to write to the cache database at the same time. A workaround is to download the grids and disable the network. Might also be worthwhile to open an issue in PROJ.

**Comment by greglucas** [comment]
I'm not sure it is due the lockfile change in this PR either (I think this has happened sporadically on some of the other PRs and even for me locally too), so I think this PR is good, modulo my comment above about the naming of "mpl".

**Pull request merged**